pkg/sql/catalog/table_elements.go

```diff
@@ -411,27 +411,85 @@ type Constraint interface {
 	// GetName returns the name of this constraint update mutation.
 	GetName() string
 
-	// NotNullColumnID returns the underlying not-null column ID, if there is one.
-	NotNullColumnID() descpb.ColumnID
-
 	// AsCheck returns the underlying check constraint, if there is one.
-	AsCheck() *descpb.TableDescriptor_CheckConstraint
+	AsCheck() CheckConstraint
 
 	// AsForeignKey returns the underlying foreign key constraint, if there is
 	// one.
-	AsForeignKey() *descpb.ForeignKeyConstraint
+	AsForeignKey() ForeignKeyConstraint
 
 	// AsUniqueWithoutIndex returns the underlying unique without index
 	// constraint, if there is one.
-	AsUniqueWithoutIndex() *descpb.UniqueWithoutIndexConstraint
+	AsUniqueWithoutIndex() UniqueWithoutIndexConstraint
 
-	// AsPrimaryKey returns the index descriptor backing the PRIMARY KEY
-	// constraint, if there is one.
-	AsPrimaryKey() Index
+	// AsUnique returns the index descriptor backing the UNIQUE or the
+	// PRIMARY KEY constraint, if there is one.
+	AsUnique() UniqueWithIndexConstraint
+}
+
+// UniqueWithIndexConstraint is an interface around a unique constraint
+// which backed by an index.
+type UniqueWithIndexConstraint interface {
+	Constraint
+	Index
+}
+
+// WithoutIndexConstraint is the supertype of all constraint subtypes which are
+// not backed by an index.
+type WithoutIndexConstraint interface {
+	Constraint
+}
+
+// CheckConstraint is an interface around a check constraint.
+type CheckConstraint interface {
+	WithoutIndexConstraint
+
+	// CheckDesc returns the underlying descriptor protobuf.
+	CheckDesc() *descpb.TableDescriptor_CheckConstraint
+
+	// Expr returns the check expression as a string.
+	Expr() string
+
+	// NumReferencedColumns returns the number of columns referenced in the check
+	// expression.
+	NumReferencedColumns() int
+
+	// GetReferencedColumnID returns the ID of the column referenced in the check
+	// expression at ordinal `columnOrdinal`.
+	GetReferencedColumnID(columnOrdinal int) descpb.ColumnID
+
+	// IsNotNullColumnConstraint returns true iff this check constraint is a
+	// NOT NULL on a column.
+	IsNotNullColumnConstraint() bool
+}
+
+// ForeignKeyConstraint is an interface around a check constraint.
+type ForeignKeyConstraint interface {
+	WithoutIndexConstraint
+
+	// ForeignKeyDesc returns the underlying descriptor protobuf.
+	ForeignKeyDesc() *descpb.ForeignKeyConstraint
+
+	// GetReferencedTableID returns the ID of the table referenced by this
+	// foreign key.
+	GetReferencedTableID() descpb.ID
```

Comment:
We probably will also need a method that retrieves referenced columns in the referenced table.

Reply:
The idea is to fall back to the descpb-protobuf only as a last resort. I'm adding these.